runtime/vm/resolvefield.cpp

```diff
@@ -578,6 +583,124 @@ addHiddenInstanceField(J9JavaVM *vm, const char *className, const char *fieldNam
+	}
+}
+
+/**
+ * Create and initialize a class' hotFieldsInfo pool element.
+ * Valid if dynamicBreadthFirstScanOrdering is enabled.
+ *
+ * @param javaVM[in] pointer to the J9JavaVM
+ * @param clazz pointer to the class where a hot field should be added
+ * @param fieldOffset value of the field offset that should be added as a hot field for the given class
+ * @param reducedCpuUtil normalized cpu utilization of the method reporting the hot field
+ * @param reducedFrequency normalized block frequency of the hot field for the method reporting the hot field
+ */
+VMINLINE void
+createClassHotFieldsInfo(J9JavaVM *javaVM, J9Class* clazz, U_8 fieldOffset, int32_t reducedCpuUtil, U_32 reducedFrequency)
+{
+	omrthread_monitor_enter(javaVM->hotFieldClassInfoPoolMutex);
+	if (NULL == clazz->hotFieldsInfo) {
+		/* initialize new hotFieldsInfo pool element as it does not exist already */
+		clazz->hotFieldsInfo = (J9ClassHotFieldsInfo *)pool_newElement(javaVM->hotFieldClassInfoPool);
+		if (NULL == clazz->hotFieldsInfo) {
+			goto exit;
+		}
+		clazz->hotFieldsInfo->hotFieldListLength = 0;
+		clazz->hotFieldsInfo->consecutiveHotFieldSelections = 0;
+		clazz->hotFieldsInfo->hotFieldOffset1 = U_8_MAX;
+		clazz->hotFieldsInfo->hotFieldOffset2 = U_8_MAX;
+		
+		/* create and initialize the new hot field */
+		omrthread_monitor_enter(clazz->classLoader->hotFieldPoolMutex);
+		J9HotField* newHotField = (J9HotField *)pool_newElement(clazz->classLoader->hotFieldPool);
+		if(NULL == newHotField) {
+			omrthread_monitor_exit(clazz->classLoader->hotFieldPoolMutex);
+			goto exit;
+		}
+		newHotField->next = NULL;
+		newHotField->hotFieldOffset = fieldOffset;
+		newHotField->hotness = (reducedFrequency * reducedCpuUtil);
+		newHotField->cpuUtil = reducedCpuUtil;
+		omrthread_monitor_exit(clazz->classLoader->hotFieldPoolMutex);
+		
+		/* update hotFieldsInfo for the class now that a new hot field has been added */
+		clazz->hotFieldsInfo->hotFieldListHead = newHotField;
+		clazz->hotFieldsInfo->isClassHotFieldListDirty = true;
+		clazz->hotFieldsInfo->hotFieldListLength++;
+		clazz->hotFieldsInfo->classLoader = clazz->classLoader;
+	}
+	
+exit:
+	omrthread_monitor_exit(javaVM->hotFieldClassInfoPoolMutex);
+}
+
+/**
+ * Add or update an existing hot field for a given class.
+ * Valid if dynamicBreadthFirstScanOrdering is enabled.
+ *
+ * @param clazz pointer to the class where a hot field should be added/updated
+ * @param fieldOffset value of the field offset that should be added/updated as a hot field for the given class
+ * @param reducedCpuUtil normalized cpu utilization of the method reporting the hot field
+ * @param reducedFrequency normalized block frequency of the hot field for the method reporting the hot field
+ */
+VMINLINE void
+addOrUpdateHotField(J9Class* clazz, U_8 fieldOffset, int32_t reducedCpuUtil, U_32 reducedFrequency)
+{
+	/* Check if the hot field exists already */
+	J9HotField* current;
+	current = clazz->hotFieldsInfo->hotFieldListHead;
```

Comment:
you still need `previous` because when you are allocating a new one you need to add it to the list

Reply:
if previous is NULL, that is a signal that you need to set the `hotFieldListHead` instead